### PR TITLE
Remove SimpleMap, Add Flatten into Block Enumeration

### PR DESCRIPTION
**Pull request by jnjosh**
@prachigauriar @macdrevx Please Review…

- [x] End the life of TWTSimpleMap
- [x] Implementation of flatten for arrays and dictionaries
- [x] Create better randomized tests
- [x] Documentation

### CODE REVIEW

File: Foundation/Block Enumeration/TWTBlockEnumeration.m [hunk 1]
```diff
@@ -75,6 +77,47 @@ + (id)performCollectOnObject:(id <NSFastEnumeration>)object resultsCollectionCla
...
++ (id)performCollectionFlattenOnObject:(id <NSFastEnumeration>)collection resultsCollectionClass:collectionClass
+{
+    NSParameterAssert(collection);
+    NSParameterAssert(collectionClass);
+    
+    id flattenedCollection = [[collectionClass alloc] init];
+
+    for (id element in collection) {
+        if ([element respondsToSelector:@selector(objectAtIndex:)]) {
+            [flattenedCollection addObjectsFromArray:[self performCollectionFlattenOnObject:element resultsCollectionClass:collectionClass]];
```
Comment (macdrevx) [inline Foundation/Block Enumeration/TWTBlockEnumeration.m]: Should the results collection class passed here always be `NSMutableArray`?

File: ToastTests/Foundation/Block Enumeration/TWTBlockEnumerationTests.m [hunk 2]
```diff
@@ -237,6 +257,27 @@ - (void)testCollectionBlockEnumerationCollect
...
+    for (NSArray *randomStringArray in randomArray) {
+        [expectedArray addObjectsFromArray:randomStringArray];
+    }
+
+    NSArray *flattenedArray = [randomArray twt_flatten];
+    
+    XCTAssertNotNil(flattenedArray, @"Returned collection is nil");
+    XCTAssertEqual([expectedArray count], [flattenedArray count], @"Returned collection does not match size of original collection");
+    XCTAssertEqualObjects(expectedArray, flattenedArray, @"Flattened collection does not match the expected collection");
+}
```
Comment (macdrevx) [inline ToastTests/Foundation/Block Enumeration/TWTBlockEnumerationTests.m]: Need tests on each of the receiving collection classes. I think trying this on anything other than `NSArray` will expose the bug I mentioned above.

File: Foundation/Block Enumeration/TWTBlockEnumeration.m (hunk 1 above)
Comment (jnjosh) [inline Foundation/Block Enumeration/TWTBlockEnumeration.m]: Updated in 85fa446

File: ToastTests/Foundation/Block Enumeration/TWTBlockEnumerationTests.m (hunk 2 above)
Comment (jnjosh) [inline ToastTests/Foundation/Block Enumeration/TWTBlockEnumerationTests.m]: Updated in 85fa446

File: Foundation/Block Enumeration/TWTBlockEnumeration.m (hunk 1 above)
Comment (macdrevx) [inline Foundation/Block Enumeration/TWTBlockEnumeration.m]: This is probably still broken for `NSOrderedSet`. Why not combine the cases like this:

```
if ([element responsdsToSelector:@selector(countByEnumeratingWithState:objects:count:)]) {
    [flattenedCollection addObjectsFromArray:[self performCollectionFlattenOnObject:element resultsCollectionClass:[NSMutableArray class]]];
}
```

Comment (macdrevx) [inline Foundation/Block Enumeration/TWTBlockEnumeration.m]: The `respondsToSelector:` check looks for the one method in the `NSFastEnumeration` protocol.

Comment (jnjosh) [inline Foundation/Block Enumeration/TWTBlockEnumeration.m]: Good catch, that makes much more sense. I've improved this and the testing for it in 903f3e5

**Comment by macdrevx** [comment]
:+1:

File: Foundation/Block Enumeration/TWTBlockEnumeration.h [hunk 3]
```diff
@@ -83,6 +83,20 @@ typedef id (^TWTBlockEnumerationInjectBlock)(id memo, id element);
...
 /*!
+ @abstract Return a newly created collection that is the result of flattening each child collection 
+    into the top top level element of a single collection.
+ @discussion Given a collection of [ [1, 2, 3], [4, 5, 6] ] this would return a collection of 
+    [1, 2, 3, 4, 5, 6]. For dictionaries, it merges child dictionaries into a single dictionary. 
+    For example, { "key1" : value1, "key2" : value2, "child" : { "key3" : value3 } } would be flattened
+    into { "key1" : value1, "key2" : value2, "key3" : value3 }. When dealing with dictionaries, this
+    is a bit more destructive and should be done with care as child keys can overwrite matching keys 
+    in the final dictionary.
+ @result A new instance of the collection with the results of flattening each element into a single
```
Comment (prachigauriar) [inline Foundation/Block Enumeration/TWTBlockEnumeration.h]: That’s not true of NSEnumerator.

Comment (jnjosh) [inline Foundation/Block Enumeration/TWTBlockEnumeration.h]: Updated in fdba5db

**Pull request merged**